// nodeunit/lib/reporters/koans.js
___ = "Fill me in!";

/*!
 * Nodeunit
 * Copyright (c) 2010 Caolan McMahon
 * MIT Licensed
 */

/**
 * Module dependencies
 */

var nodeunit = require('../nodeunit'),
    utils = require('../utils'),
    fs = require('fs'),
    track = require('../track'),
    path = require('path'),
    AssertionError = require('../assert').AssertionError;

/**
 * Reporter info string
 */

exports.info = "A more philosophical test runner";


/**
 * Run all tests within each module, reporting the results to the command-line.
 *
 * @param {Array} files
 * @api public
 */

exports.run = function (files, options, callback) {

    if (!options) {
        // load default options
        var content = fs.readFileSync(
            __dirname + '/../../bin/nodeunit.json', 'utf8'
        );
        options = JSON.parse(content);
    }

    var error = function (str) {
        return options.error_prefix + str + options.error_suffix;
    };
    var ok    = function (str) {
        return options.ok_prefix + str + options.ok_suffix;
    };
    var bold  = function (str) {
        return options.bold_prefix + str + options.bold_suffix;
    };
    var assertion_message = function (str) {
        return options.assertion_prefix + str + options.assertion_suffix;
    };

    var start = new Date().getTime();
    var tracker = track.createTracker(function (tracker) {
        if (tracker.unfinished()) {
            console.log('');
            console.log(error(bold(
                'FAILURES: Undone tests (or their setups/teardowns): '
            )));
            var names = tracker.names();
            for (var i = 0; i < names.length; i += 1) {
                console.log('- ' + names[i]);
            }
            console.log('');
            console.log('To fix this, make sure all tests call test.done()');
            process.reallyExit(tracker.unfinished());
        }
    });

    var wisdom = ['Never leave rollerskates at the top of the stairs',
    'Never play leapfrog with a unicorn', 'Its not JavaScript its what you do with it that counts',
    'Bird never lands in clenched fist', 'Dynamic languages never change',
    'Dont get mad get even', 'JavaScript wasnt built in a day- it was a weekend',
    'Never use == when === will do', '"2" + 2 always makes a 22 in JavaScript!'
    ];

    var person = ['Ninja', 'Guru', 'Master', 'Deity', 'Sensei', 'Splinter'];

    var fileName;

    var opts = {
        testspec: options.testspec,
        testFullSpec: options.testFullSpec,
        moduleStart: function (name) {
            fileName = name;
            console.log('\n' + bold(name));
        },
        testDone: function (name, assertions) {
            tracker.remove(name);

            if (!assertions.failures()) {
                console.log('✔ ' + name);
            }
            else {
                assertions.forEach(function (a) {
                    if (a.failed()) {
                        console.log(error(bold('✖ ' + fileName + "#" + name) + " has damaged your karma.\n"));

                        a = utils.betterErrors(a);
                        if (a.error instanceof AssertionError && a.message) {
                            console.log(
                                'Assertion Message: ' +
                                assertion_message(a.message)
                            );
                        }

                        var backtrace = a.error.stack.split("\n");
                        console.log(backtrace[0]);

                        for(var i = 1; i < backtrace.length; i++) {
                            if(backtrace[i].match(/about_/)) {
                                console.log(backtrace[i] + '\n');
                            }
                        }
                    }
                });
            }
        },
        done: function (assertions, end) {
            var end = end || new Date().getTime();
            var duration = end - start;
            if (assertions.failures()) {
                console.log("\nThe Master says:");
                console.log("You have not yet reached enlightenment.");
                console.log("Do not lose hope.\n");
                console.log("Remember: ");
                console.log(wisdom[Math.floor(Math.random() * wisdom.length)]);
                console.log("\nThe answers you seek:");

                console.log(
                    '\n' + bold(error('FAILURES: ')) + assertions.failures() +
                    '/' + assertions.length + ' assertions failed (' +
                    assertions.duration + 'ms)'
                );
            }
            else {
                console.log(
                   '\n' + bold(ok('OK: ')) + assertions.length +
                   ' assertions (' + assertions.duration + 'ms)'
                );
            }

            if (callback) callback(assertions.failures() ? new Error('We have got test failures.') : undefined);
        },
        testStart: function(name) {
            tracker.put(name);
        }
    };
    if (files && files.length) {
        var paths = files.map(function (p) {
            return path.resolve(p);
        });
        nodeunit.runFiles(paths, opts);
    } else {
        nodeunit.runModules(files,opts);
    }
};
